### DISCUSSION TITLE
Get the last NN layer from the deployed model

### DISCUSSION THREAD

**Discussion by ShuaihuaLu** [mir-group/nequip#187] [Q&A] [answered]
![Picture1](https://user-images.githubusercontent.com/84307614/161072427-fa3b2d12-3883-49e7-9302-4683568e6bb0.png)
Can we get the last layer from the deployed model after model training? If so, how to do that?

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @ShuaihuaLu ,

1) what do you mean "get the last layer"?

2) that looks like a saved dataset not a deployed model in your screenshot?

**Reply by ShuaihuaLu** [reply]
Thank you for your reply. After training the nequip and getting the deployed model. I want to get the output of the last layer before the predicted values because we want to see the internal representation and use UMAP to analysis them.  I  have the best-model.pth which did not shown in the screenshot I put.

**Reply by Linux-cpp-lisp** [reply]
Hi @ShuaihuaLu ,

I see— the output of the last hidden feature layer is still accessible in the output dict of the model in the `node_features` key. If you need the features from an earlier part of the model, you have to add some way to save those values for later, see https://github.com/mir-group/nequip/blob/develop/examples/monkeypatch.py for a sketch of how you might do that.

But generally it sounds like you want the final `node_features`; for this I would use the `nequip-evaluate` command to evaluate your model on the data and use the `--output file.xyz` and `--output-fields node_features` to get `file.xyz` with the predicted features on a dataset.

**Reply by ShuaihuaLu** [reply]
Hi @Linux-cpp-lisp,
Thanks a lot for your guidance and suggestion. I did what you suggested using the commend 'nequip-evaluate --train-dir results/toluene/example-run-toluene/ --output file.xyz --output-fields node_features'. I got a .xyz file and it contains node features. Are those features the weights of last layer before output ? Besides, I did not understand the means of features. Could you explain it for me, please?

**Reply by Linux-cpp-lisp** [reply]
No, the output `node_features` are the predicted _values_ for the features of each atom in those structures. The actual weights of the last self-interaction layer can be read from the model file itself. The NequIP paper on the arXiv covers the architecture of the network in more detail.

**Reply by ShuaihuaLu** [reply]
I see. I think it has solved my problem. Thank you so much for your help.
